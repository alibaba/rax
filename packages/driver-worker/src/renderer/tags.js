const customTags = {
  audio: true,
  button: true,
  canvas: true,
  checkbox: true,
  'checkbox-group': true,
  icon: true,
  label: true,
  picker: true,
  'picker-view': true,
  'picker-view-column': true,
  radio: true,
  slider: true,
  text: true,
  video: true,
  image: true,
  input: true,
  map: true,
  'radio-group': true,
  swiper: true,
  'swiper-item': true,
  textarea: true,
  view: true,
  form: true,
  navigator: true,
  progress: true,
  'scroll-view': true,
  switch: true,
  'web-view': true,
  page: true
};

const w3cTags = {
  style: true
};

export function isValidCustomTags(tagName) {
  return customTags.hasOwnProperty(tagName);
}

export function isValidW3CTag(tagName) {
  return w3cTags.hasOwnProperty(tagName);
}
