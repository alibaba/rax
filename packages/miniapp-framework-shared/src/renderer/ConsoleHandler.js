import retrocycle from '../retrocycle';

export function createConsoleHandler() {
  return function apply(evt) {
  };
}
