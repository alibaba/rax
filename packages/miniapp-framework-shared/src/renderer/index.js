import { setupAppear } from 'runtime-shared/src/appear';
import { setupTap } from './tap';
import { setupTheme } from './theme';
import { setupReachBottom } from './reachBottom';

export {
  setupAppear,
  setupTap,
  setupTheme,
  setupReachBottom,
};
