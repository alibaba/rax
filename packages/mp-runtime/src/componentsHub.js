const components = {};

export function registerComponent(path, component) {
  components[path] = component;
}

export function getComponent(path) {
  return components[path];
}
