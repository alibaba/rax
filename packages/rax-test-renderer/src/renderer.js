import { shared } from 'rax';
import ServerDriver from 'driver-server';
import Serializer from './serializer';

const { Instance, Host } = shared;

// For inject Host init options
Host.driver = ServerDriver;

export default {
  create(element) {
    let container = ServerDriver.createBody();
    let rootInstance = Instance.mount(element, container, {});
    // The field corresponding to node and instance is _r, which you can see in rax/src/vdom/instance
    let renderedComponent = rootInstance.__getRenderedComponent();

    renderedComponent.toJSON = () => {
      return new Serializer(container).toJSON();
    };

    renderedComponent.getInstance = () => {
      return renderedComponent._instance;
    };

    renderedComponent.update = (element) => {
      Instance.mount(element, container, {});
    };

    renderedComponent.unmount = () => {
      let component = Instance.get(container);

      if (!component) {
        return false;
      }

      Instance.remove(container);
      component._internal.unmountComponent();

      return true;
    };

    return renderedComponent;
  }
};
