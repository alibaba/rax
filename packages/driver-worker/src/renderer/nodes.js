import { isValidW3CTag, isValidCustomTags } from './tags';


const CUSTOM_TAG_PREFIX = 'a-';

/**
 * Only node name that in whitelist can be created.
 */
export function getTagName(nodeName) {
  const tagName = String(nodeName).toLowerCase();
  if (isValidW3CTag(tagName)) {
    return tagName;
  } else if (isValidCustomTags(tagName)) {
    return CUSTOM_TAG_PREFIX + tagName;
  } else {
    return 'a-unknown';
  }
}

export function createNode(vnode) {
  let node, tagName;
  if (vnode.nodeType === 3) {
    node = document.createTextNode(vnode.data);
  } else if (vnode.nodeType === 1 && (tagName = getTagName(vnode.nodeName))) {
    node = document.createElement(tagName);
  } else if (vnode.nodeType === 8) {
    node = document.createComment(vnode.data);
  }

  return node;
}
