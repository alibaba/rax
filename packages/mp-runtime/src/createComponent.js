import computeChangedData from './computeChangedData';
import deepCopy from './deepCopy';
import { registerComponent } from './componentsHub';
import { normalizeScopedSlots } from './normalizeScopedSlots';
import createLifecycleAdapter from './lifecycleAdapter';

const STYLE_FLAG_KEY = '_appendedComponentStyle';

/**
 * Returns a boolean indicating whether the object has the specified property as its own property.
 * @param {Object} object
 * @param {String} property
 */
function hasOwnProperty(object, property) {
  return Object.prototype.hasOwnProperty.call(object, property);
}

function getSlotName(item) {
  if (item && hasOwnProperty(item, 'props')) {
    return item.props.slot || 'default';
  } else {
    return 'default';
  }
}

function injectSlot(child, $slots) {
  if (null == child) return;
  const slotName = getSlotName(child);
  if (null == slotName) return;
  $slots[slotName] = $slots[slotName] || [];
  $slots[slotName].push(child);
}

// Count of component instance numbers.
let componentCount = 0;

export default function createComponent(renderFactory, render, config, componentPath, cssText, type = 'ali') {
  const templateRender = renderFactory(render);
  let props = {};
  switch (type) {
    case 'weixin':
      if (config.properties && typeof config.properties === 'object') {
        Object.keys(config.properties).forEach(property => {
          const propValue = config.properties[property];
          if (typeof propValue === 'object' && propValue.value !== undefined) {
            props[property] = propValue.value;
          }
        });
      }
      break;
    case 'ali':
    default:
      props = config.props;
      break;
  }
  const component = class extends render.Component {
    static contextTypes = {
      $page: null,
    };

    constructor(props, context) {
      super(props, context);
      /**
       * One component's style can only be append once in each docuemnt(page).
       * Use a flag in document object to mark.
       */
      if (context.$page && context.$page.vnode) {
        const document = context.$page.vnode._document;
        const styleFlag = document[STYLE_FLAG_KEY] = document[STYLE_FLAG_KEY] || {};
        if (!styleFlag[componentPath] && cssText) {
          styleFlag[componentPath] = true;
          const cssTextNode = document.createTextNode(String(cssText));
          const styleNode = document.createElement('style');
          styleNode.appendChild(cssTextNode);
          document.body.appendChild(styleNode);
        }
      }

      /**
       * If not defined `data` field in config,
       * then default val will be an empty plain object,
       * else data will be deep copied.
       */
      this.state = hasOwnProperty(config, 'data')
        ? deepCopy(config.data)
        : {};
      this.publicInstance = this._createPublicInstance();
      this.componentId = ++componentCount;
      this.lifecycleAdapter = createLifecycleAdapter(this.publicInstance, config, type);
      this.lifecycleAdapter._constructor();
    }

    static defaultProps = props;

    _createPublicInstance() {
      const scope = {};

      if (config.methods != null) {
        for (let key in config.methods) {
          if (hasOwnProperty(config.methods, key)) {
            scope[key] = config.methods[key].bind(scope);
          }
        }
      }

      Object.defineProperty(scope, 'props', {
        get: () => this.props,
      });
      Object.defineProperty(scope, 'data', {
        get: () => this.state,
      });

      Object.defineProperty(scope, 'setData', {
        get: () => this.setData,
      });

      Object.defineProperty(scope, '$slots', {
        get: () => this.transformChildrenToSlots(this.props.children),
      });

      Object.defineProperty(scope, '$scopedSlots', {
        get: () => normalizeScopedSlots(this.props.scopedSlots, scope.$slots),
      });

      Object.defineProperty(scope, 'is', {
        get: () => componentPath,
      });

      Object.defineProperty(scope, '$page', {
        get: () => this.context.$page,
      });

      Object.defineProperty(scope, '$id', {
        get: () => this.componentId,
      });

      return scope;
    }

    setData = (data, callback) => {
      if (data == null) return;
      this.setState(computeChangedData(this.state, data), callback);
    };

    transformChildrenToSlots = (children) => {
      const $slots = {};
      if (Array.isArray(children)) {
        for (let i = 0, l = children.length; i < l; i++) {
          injectSlot(children[i], $slots);
        }
      } else {
        injectSlot(children, $slots);
      }
      return $slots;
    };

    componentWillMount() {
      if (this.lifecycleAdapter.componentWillMount) {
        this.lifecycleAdapter.componentWillMount(this.props);
      }
    }

    componentDidMount() {
      if (this.lifecycleAdapter.componentDidMount) {
        this.lifecycleAdapter.componentDidMount();
      }
    }

    componentWillUpdate(nextProps) {
      if (this.lifecycleAdapter.componentWillUpdate) {
        this.lifecycleAdapter.componentWillUpdate(nextProps);
      }
    }

    componentDidUpdate(prevProps, prevState) {
      if (this.lifecycleAdapter.componentDidUpdate) {
        this.lifecycleAdapter.componentDidUpdate(prevProps, prevState);
      }
    }

    componentWillReceiveProps(nextProps) {
      if (this.lifecycleAdapter.componentWillReceiveProps) {
        this.lifecycleAdapter.componentWillReceiveProps(nextProps);
      }
    }

    componentWillUnmount() {
      if (this.lifecycleAdapter.componentWillUnmount) {
        this.lifecycleAdapter.componentWillUnmount();
      }
    }

    render() {
      const { $slots, $id, props, data } = this.publicInstance;
      return templateRender.call(this.publicInstance, {
        $id, $slots, ...props, ...data
      });
    }
  };

  if (componentPath !== undefined) {
    registerComponent(componentPath, component);
  }

  return component;
}
