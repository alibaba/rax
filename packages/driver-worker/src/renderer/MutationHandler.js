import { sharedNodeMap } from './NodeMap';
import { createNode } from './nodes';
import { addEvent, removeEvent, setPostMessage } from './events';
import { setAttribute } from './attrs';

const TEXT_CONTENT = 'textContent';
const TEXT_CONTENT_ATTR = TEXT_CONTENT in document ? TEXT_CONTENT : 'nodeValue';

/**
 * Returns "attributes" if it was an attribute mutation.
 * "characterData" if it was a mutation to a CharacterData node.
 * And "childList" if it was a mutation to the tree of nodes.
 */
export default class MutationHandler {
  constructor(sender) {
    setPostMessage(sender);
  }

  apply(data) {
    let mutations = data.mutations;
    for (let i = 0; i < mutations.length; i++) {
      // apply mutation
      let mutation = mutations[i];
      this[mutation.type](mutation);
    }
  }

  childList({ target, removedNodes, addedNodes, nextSibling }) {
    let vnode = target;

    if (vnode && vnode.nodeName === 'BODY') {
      document.body.$$id = vnode.$$id;
    }

    let parent = sharedNodeMap.get(vnode);
    if (removedNodes) {
      for (let i = removedNodes.length; i--;) {
        let node = sharedNodeMap.get(removedNodes[i]);
        sharedNodeMap.delete(node);
        if (parent && node) {
          parent.removeChild(node);
        }
      }
    }

    if (addedNodes) {
      for (let i = 0; i < addedNodes.length; i++) {
        let newNode = sharedNodeMap.get(addedNodes[i]);
        if (!newNode) {
          newNode = createNode(addedNodes[i]);
        }

        if (parent) {
          parent.insertBefore(newNode, nextSibling && sharedNodeMap.get(nextSibling) || null);
        }
      }
    }
  }

  attributes({ target, attributeName, newValue, style }) {
    let node = sharedNodeMap.get(target);
    // Node maybe null when node is removed and there is a setInterval change the node that will cause error
    if (!node) return;

    if (style) {
      setAttribute(node, 'style', style);
    }
    if (attributeName) {
      setAttribute(node, attributeName, newValue);
    }
  }

  characterData({ target, newValue }) {
    let node = sharedNodeMap.get(target);
    node[TEXT_CONTENT_ATTR] = newValue;
  }

  addEvent({ target, eventName }) {
    let node = sharedNodeMap.get(target);
    if (!node) return;

    addEvent(node, eventName);
  }

  removeEvent({ target, eventName }) {
    let node = sharedNodeMap.get(target);
    if (!node) return;

    removeEvent(node, eventName);
  }

  canvasRenderingContext2D({ target, method, args, properties }) {
    let canvas = sharedNodeMap.get(target);
    if (!canvas) return;

    let context = canvas.getContext('2d');

    if (properties) {
      for (let key in properties) {
        if (properties.hasOwnProperty(key)) {
          context[key] = properties[key];
        }
      }
    }

    if (method) {
      context[method].apply(context, args);
    }
  }

  lottieRenderingContext({ target, method, args }) {
    let lottie = sharedNodeMap.get(target);

    if (method) {
      lottie[method].apply(lottie, args);
    }
  }
}
