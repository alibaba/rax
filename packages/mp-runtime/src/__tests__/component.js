import Rax, { createElement } from 'rax';
import renderer from 'rax-test-renderer';
import createComponent from '../createComponent';
import { getComponent } from '../componentsHub';
import createWorkerDriver from 'driver-worker';

describe('Mini Program Component', () => {
  it('should render default slots', () => {
    function renderFactory(Rax) {
      return function(instance, data) {
        return (
          <view>Hello {this.$slots.default}</view>
        );
      };
    }
    const Comp = createComponent(renderFactory, Rax, {});
    const tree = renderer.create(
      <Comp page="https://example.com/">Example</Comp>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render data and props', () => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {
          foo: data.foo,
          bar: data.bar,
        });
      };
    }
    const Comp = createComponent(renderFactory, Rax, {
      data: { foo: 'from data' },
    });
    const tree = renderer.create(createElement(Comp, { bar: 'from prop' }));
    expect(tree).toMatchSnapshot();
  });

  it('should works well with didMount lifecycle', () => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {
          foo: data.foo,
        });
      };
    }
    const Comp = createComponent(renderFactory, Rax, {
      data: { foo: 'from data' },
      didMount() {
        this.setData({ foo: 'modified in didMount' });
      },
    });
    const tree = renderer.create(createElement(Comp));
    expect(tree).toMatchSnapshot();
  });

  it('should works well with didUpdate and didUnmount lifecycle', (done) => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {
          foo: data.foo,
        });
      };
    }
    const Comp = createComponent(renderFactory, Rax, {
      data: { foo: 'from data' },
      didMount() {
        this.setData({ foo: 'modified in didMount' });
      },
      didUpdate(prevProps, prevData) {
        expect(prevProps).toEqual({ bar: 1 });
        expect(prevData).toEqual({ foo: 'from data' });
      },
      didUnmount() {
        // if did unmount not triggered, task will not be done.
        done();
      },
    });
    const tree = renderer.create(createElement(Comp, { bar: 1 }));
    tree.unmount();
  });

  it('should works well with setData', (done) => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {
          foo: data.foo,
        });
      };
    }
    const Comp = createComponent(renderFactory, Rax, {
      data: { arr: [{ a: 1, b: 2 }] },
      didMount() {
        this.setData({
          'arr[0].a': 3,
          'arr[2]': { a: 10, b: 20 }
        });
        expect(this.data).toEqual({
          arr: [
            { a: 3, b: 2 },
            undefined,
            { a: 10, b: 20 },
          ]
        });
      },
      didUnmount() {
        done();
      },
    });
    const tree = renderer.create(createElement(Comp, { bar: 1 }));
    tree.unmount();
  });

  it('should works well with named slot', () => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {}, [data.$slots.default, data.$slots.named]);
      };
    }
    const Comp = createComponent(renderFactory, Rax, { data: {} });
    const tree = renderer.create(createElement(Comp, {}, [
      'child',
      createElement('text', { slot: 'named' }, 'named slot')
    ]));

    expect(tree).toMatchSnapshot();
  });

  it('should works well with methods', (done) => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {}, data.count);
      };
    }
    const Comp = createComponent(renderFactory, Rax, {
      data: { count: 0 },
      methods: {
        countAdd() {
          this.setData({ count: this.data.count + 1 });
        },
      },
      didMount() {
        this.countAdd();
      },
      didUnmount() {
        done();
      },
    });
    const tree = renderer.create(createElement(Comp, {}));
    expect(tree.toJSON()).toMatchSnapshot();
    tree.unmount();
  });

  it('should works well with register components', () => {
    const componentPath = 'path/to/component';
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', { foo: 'bar' });
      };
    }

    const Comp = createComponent(renderFactory, Rax, {}, componentPath);
    expect(getComponent(componentPath)).toEqual(Comp);
  });

  it('should get componet is, $page, $id', (done) => {
    const $page = Symbol('$page');

    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {}, data.count);
      };
    }
    const compPath = 'path/to/component';
    const Comp = createComponent(renderFactory, Rax, {
      data: { count: 0 },
      didMount() {
        expect(this.is).toEqual(compPath);
        expect(typeof this.$id).toEqual('number');
        expect(this.$page).toEqual($page);
        done();
      },
    }, compPath);

    class Page extends Rax.Component {
      static contextTypes = { $page: null };

      getChildContext() {
        return { $page };
      }

      render() {
        return createElement('view', {}, [createElement(Comp)]);
      }
    }

    renderer.create(createElement(Page));
  });

  it('should have default val for data', (done) => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {
          foo: data.foo,
        });
      };
    }
    const Comp = createComponent(renderFactory, Rax, {
      // data field is empty.
      didMount() {
        this.someMethod();
        done();
      },
      methods: {
        someMethod() {
          expect(this.data).toEqual({});
        },
      }
    });

    renderer.create(createElement(Comp, { bar: 'from prop' }));
  });

  it('should render style once', (done) => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {
          className: 'foo',
        });
      };
    }
    const cssText = '.foo { color: red; }';
    const Comp = createComponent(renderFactory, Rax, {}, 'path/to/component', cssText);

    global.setImmediate = null;
    const driver = createWorkerDriver({
      postMessage(message) {
        expect(message).toMatchSnapshot();
        done();
      },
      addEventListener() {},
    });

    class Page extends Rax.Component {
      static contextTypes = { $page: null };

      getChildContext() {
        const $page = {
          vnode: { _document: driver.document }
        };
        return { $page };
      }

      render() {
        // Render Comp 4 times.
        return createElement('view', null, [
          createElement(Comp),
          createElement(Comp),
          createElement(Comp),
          createElement(Comp),
        ]);
      }
    }

    Rax.render(createElement(Page), null, { driver });
  });

  it('should render weixin component props', () => {
    function renderFactory(Rax) {
      return function(data) {
        return createElement('view', {
          bar: data.bar,
        });
      };
    }
    const Comp = createComponent(renderFactory, Rax, {
      properties: {
        bar: {
          type: String,
          value: 'from prop'
        }
      },
    }, null, null, 'weixin');
    const tree = renderer.create(createElement(Comp));
    expect(tree).toMatchSnapshot();
  });
});

