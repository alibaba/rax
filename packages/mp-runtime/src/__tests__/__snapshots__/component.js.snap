// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['Mini Program Component should render data and props 1'] = `
Object {
  "attributes": Object {
    "bar": "from prop",
    "foo": "from data",
  },
  "tagName": "VIEW",
}
`;

exports['Mini Program Component should render default slots 1'] = `
Object {
  "children": Array [
    "Hello ",
    "Example",
  ],
  "tagName": "VIEW",
}
`;

exports['Mini Program Component should render style once 1'] = `
Object {
  "mutations": Array [
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "2",
          "nodeName": "STYLE",
          "nodeType": 1,
        },
      ],
      "target": Object {
        "$$id": "3",
        "nodeName": "BODY",
      },
      "type": "childList",
    },
    Object {
      "addedNodes": Array [
        Object {
          "$$id": "1",
          "data": ".foo { color: red; }",
          "nodeType": 3,
        },
      ],
      "target": Object {
        "$$id": "2",
      },
      "type": "childList",
    },
  ],
  "type": "MutationRecord",
}
`;

exports['Mini Program Component should works well with didMount lifecycle 1'] = `
Object {
  "attributes": Object {
    "foo": "modified in didMount",
  },
  "tagName": "VIEW",
}
`;

exports['Mini Program Component should works well with methods 1'] = `
Object {
  "children": Array [
    "1",
  ],
  "tagName": "VIEW",
}
`;

exports['Mini Program Component should works well with named slot 1'] = `
Object {
  "children": Array [
    "child",
    Object {
      "attributes": Object {
        "slot": "named",
      },
      "children": Array [
        "named slot",
      ],
      "tagName": "TEXT",
    },
  ],
  "tagName": "VIEW",
}
`;

exports['Mini Program Component should render weixin component props 1'] = `
Object {
  "attributes": Object {
    "bar": "from prop",
  },
  "tagName": "VIEW",
}
`;
